letsencrypt_auto/letsencrypt-auto.template

```diff
@@ -141,65 +81,146 @@ then
+# `su`
+if test "`id -u`" -ne "0" ; then
+  if command -v sudo 1>/dev/null 2>&1; then
+    SUDO=sudo
   else
-    virtualenv --no-site-packages --python $LE_PYTHON $VENV_PATH > /dev/null
+    echo \"sudo\" is not available, will use \"su\" for installation steps...
+    # Because the parameters in `su -c` has to be a string,
+    # we need properly escape it
+    su_sudo() {
+      args=""
+      # This `while` loop iterates over all parameters given to this function.
+      # For each parameter, all `'` will be replace by `'"'"'`, and the escaped string
+      # will be wrap in a pair of `'`, then append to `$args` string
+      # For example, `echo "It's only 1\$\!"` will be escaped to:
+      #   'echo' 'It'"'"'s only 1$!'
+      #     │       │└┼┘│
+      #     │       │ │ └── `'s only 1$!'` the literal string
+      #     │       │ └── `\"'\"` is a single quote (as a string)
+      #     │       └── `'It'`, to be concatenated with the strings followed it
+      #     └── `echo` wrapped in a pair of `'`, it's totally fine for the shell command itself
+      while [ $# -ne 0 ]; do
+        args="$args'$(printf "%s" "$1" | sed -e "s/'/'\"'\"'/g")' "
+        shift
+      done
+      su root -c "$args"
+    }
+    SUDO=su_sudo
   fi
 else
-  DeterminePythonVersion
+  SUDO=
 fi
 
+if [ "$1" = "--os-packages-only" ]; then
+  Bootstrap
+elif [ "$1" != "--_skip-to-install" ]; then
+  echo "Upgrading letsencrypt-auto..."
 
-printf "Updating letsencrypt and virtual environment dependencies..."
-if [ "$VERBOSE" = 1 ]  ; then
-  echo
-  $VENV_BIN/pip install -U setuptools
-  $VENV_BIN/pip install -U pip
-  $VENV_BIN/pip install -r "$LEA_PATH"/py26reqs.txt -U letsencrypt letsencrypt-apache
-  # nginx is buggy / disabled for now, but upgrade it if the user has
-  # installed it manually
-  if $VENV_BIN/pip freeze | grep -q letsencrypt-nginx ; then
-    $VENV_BIN/pip install -U letsencrypt letsencrypt-nginx
+  if [ ! -f $VENV_BIN/letsencrypt ]; then
+    OLD_VERSION="0.0.0"  # ($VENV_BIN/letsencrypt --version)
+  else
+    OLD_VERSION="0.0.0"
   fi
-else
-  $VENV_BIN/pip install -U setuptools > /dev/null
-  printf .
-  $VENV_BIN/pip install -U pip > /dev/null
-  printf .
-  # nginx is buggy / disabled for now...
-  $VENV_BIN/pip install -r "$LEA_PATH"/py26reqs.txt > /dev/null
-  printf .
-  $VENV_BIN/pip install -U letsencrypt > /dev/null
-  printf .
-  $VENV_BIN/pip install -U letsencrypt-apache > /dev/null
-  if $VENV_BIN/pip freeze | grep -q letsencrypt-nginx ; then
-    printf .
-    $VENV_BIN/pip install -U letsencrypt-nginx > /dev/null
+
+  # TODO: Don't bother upgrading if we're already up to date.
+  if [ "$OLD_VERSION" != "1.2.3" ]; then
+    Bootstrap
+    echo "Creating virtual environment..."
+    rm -rf "$VENV_PATH"
+    DeterminePythonVersion
+    if [ "$VERBOSE" = 1 ] ; then
+      virtualenv --no-site-packages --python $LE_PYTHON $VENV_PATH
+    else
+      virtualenv --no-site-packages --python $LE_PYTHON $VENV_PATH > /dev/null
+    fi
+
```

Comment:
Perhaps add a big visual separator line here, because we're about to change languages and universes :)